### ISSUE TITLE
Add token Auth

### ISSUE THREAD

**Issue by 6543** [woodpecker-ci/woodpecker#665]
Add option for token generation to auth.

So you can do more ganular access control

**Comment by anbraten** [comment]
link #368

**Comment by ambroisie** [comment]
Noting something which was discussed on chat: this would be useful for writing a NixOS test and/or other kind of integration testing.

**Comment by gzm0** [comment]
We need this and I would like to contribute this. I have sketched out a rough plan below.

I'd appreciate some early feedback from the maintainers of whether the plan looks in line with what you had in mind.

## 1. Support multiple tokens per user

- Tokens are still personal and full scope.
- Should be revokeable individually.

**Open Question**
How to build token revocation? Currently the user hash is reset to revoke the token (this even revokes the user's session tokens).
This would not work for individual token revocation. Should we just build a `token` table and a `token_id` in the JWT that we check for revocation? (we'll need a `token` table anyways so we can display the tokens to the user).

## 2. Support scoped tokens

Restrict the scope of tokens. Probably on 2 axes:
- resource (repositories or organizations)
- action (probably: admin, pull, push for starters, can add more over time for more granularity).

For backwards compatibility, missing scope value in JWT means "unrestricted".

For user tokens, we need to check both the upstream permissions from the forge and the token permissions.

## 3. Add machine accounts

- Machine accounts are never tied to a forge.
- Machine accounts cannot log in.
- Admins can create tokens for machine accounts (probably we should force explicit action scope selection).
- Permission checks are solely based on the token permissions.

**Comment by qwerty287** [comment]
I'm against adding something like machine accounts but what I could imagine is adding something like repo tokens instead, i.e. tokens linked to a single repo, but that requires refactoring the whole auth system.

Options 1 (which can be extended by 2 later) is probably more straightforward to add.

**Comment by gzm0** [comment]
Oh, sorry. Maybe I wasn't clear. The numbers in my original comment were not meant to be options, but steps.

The primary need for us is to authorize the autoscaler. It needs system wide permissions to manage system wide agents.

Currently I'm using my personal token, but in an organizational context, that's not ideal. If my user account gets disabled, autoscaling will fail.

The concept of "machine account" really is meant to serve as a "vessel" for tokens without user association (certain things track owners / creators, it feels easier to just create a "user" that tracks this metadata).

Hope that clarifies.

**Comment by qwerty287** [comment]
Ok, so yes I agree that your points 1 and 2 are good ideas.

About point 3, wouldn't something like org/repo tokens also fix your problem? Adding own "machine" users is a bad idea isln my opinion. But that could also be discussed when we're at that point

**Comment by gzm0** [comment]
> About point 3, wouldn't something like org/repo tokens also fix your problem?

Yes, they would (or system tokens more specifically, but I guess we'd introduce repo/org/system level tokens).

In fact, I've tried out this approach briefly by adding a token type ([here](https://github.com/woodpecker-ci/woodpecker/blob/7c51ff768c739ac59f61f42eb3c289256ad629e3/shared/token/token.go#L30-L37)).

Where I got "stuck" was operations where the user is taken from the token and recorded as owner like here:

https://github.com/woodpecker-ci/woodpecker/blob/7c51ff768c739ac59f61f42eb3c289256ad629e3/server/api/agent.go#L175-L189

It felt like not recording an owner is not acceptable, so essentially, we'd record the token as an owner. At that point, it felt easier to just introduce robot accounts.

But maybe it is better to introduce this additional token type to make it crystal clear how authorization works:

- UserToken: identifies the user, authorization by the forge, if we build (2) privileges can be limited (but *not* extended) via token scope.
- BotToken: no user association, authorization by association (org, repo, system) & privilege limits at token creation time.

The downside of such a solution seems to be that it has a more profound impact on the data-model: every "owner' field will need to accept either a user or a bot token.

Another option is that we model this differently in the DB: A Bot token gets a row in the `user` table but we display these differently. My fear with this is that it will make multiple tokens per user extremely messy.

WDYT?

**Comment by qwerty287** [comment]
I don't think we need the ownership of agents etc. at all. We can replace that quite easily by a different solution. The only line of code where the `OwnerID` of agents is used seems to be https://github.com/woodpecker-ci/woodpecker/blob/main/server/model/agent.go#L55 - but if we add a system token, all agents added with system tokens can just get `OwnerID = IDNotSet` and if we have org tokens, they also just get `OwnerID = IDNotSet` but set an `OrgID`. We have to make sure of course that this has no side effects, but I don't think so. If owner IDs are stored at more points this can likely be solved similarly.

**Comment by gzm0** [comment]
Huh, indeed. So yeah, then I agree, bot tokens are a better idea. If you feel this is worthwhile building, I'll start :)

**Comment by qwerty287** [comment]
Nice - yes, go on please! Maybe not everything in one PR, but you could start with step 1 (multiple tokens per user), then once this is done/merged step 2 and finally 3

**Comment by gzm0** [comment]
> Maybe not everything in one PR

Yes, for sure not! I'll start hacking on this and see what kind of split makes most sense.
